mm2src/coins/qrc20/swap_ops.rs

```diff
@@ -0,0 +1,882 @@
+#[derive(Debug)]
+pub struct ReceiverSpendDetails {
+    pub swap_id: Vec<u8>,
+    pub value: U256,
+    pub secret: Vec<u8>,
+    pub token_address: H160,
+    pub sender: H160,
+}
+
+impl Qrc20Coin {
+    pub async fn send_hash_time_locked_payment(
+        &self,
+        id: Vec<u8>,
+        value: U256,
+        time_lock: u32,
+        secret_hash: Vec<u8>,
+        receiver_addr: H160,
+    ) -> Result<TransactionEnum, String> {
+        let allowance = try_s!(self.allowance(self.swap_contract_address).await);
+
+        let mut outputs = Vec::default();
+        // check if we should reset the allowance to 0 and raise this to the max available value (our balance)
+        if allowance < value {
+            let balance = try_s!(self.my_balance().compat().await);
+            let balance = try_s!(wei_from_big_decimal(&balance, self.utxo.decimals));
+            if allowance > U256::zero() {
+                // first reset the allowance to the 0
+                outputs.push(try_s!(self.approve_output(self.swap_contract_address, 0.into())));
+            }
+            // set the allowance from 0 to `balance` after the previous output will be executed
+            outputs.push(try_s!(self.approve_output(self.swap_contract_address, balance)));
+        }
+
+        // when this output is executed, the allowance will be sufficient already
+        outputs.push(try_s!(self.erc20_payment_output(
+            id,
+            value,
+            time_lock,
+            &secret_hash,
+            receiver_addr
+        )));
+
+        self.send_contract_calls(outputs).await
+    }
+
+    pub async fn spend_hash_time_locked_payment(
+        &self,
+        payment_tx: UtxoTx,
+        secret: Vec<u8>,
+    ) -> Result<TransactionEnum, String> {
+        let Erc20PaymentDetails {
+            swap_id, value, sender, ..
+        } = try_s!(self.erc20_payment_details_from_tx(&payment_tx).await);
+
+        let status = try_s!(self.payment_status(swap_id.clone()).await);
+        if status != eth::PAYMENT_STATE_SENT.into() {
+            return ERR!("Payment state is not PAYMENT_STATE_SENT, got {}", status);
+        }
+
+        let spend_output = try_s!(self.receiver_spend_output(swap_id, value, secret, sender));
+        self.send_contract_calls(vec![spend_output]).await
+    }
+
+    pub async fn refund_hash_time_locked_payment(&self, payment_tx: UtxoTx) -> Result<TransactionEnum, String> {
+        let Erc20PaymentDetails {
+            swap_id,
+            value,
+            receiver,
+            secret_hash,
+            ..
+        } = try_s!(self.erc20_payment_details_from_tx(&payment_tx).await);
+
+        let status = try_s!(self.payment_status(swap_id.clone()).await);
+        if status != eth::PAYMENT_STATE_SENT.into() {
+            return ERR!("Payment state is not PAYMENT_STATE_SENT, got {}", status);
+        }
+
+        let refund_output = try_s!(self.sender_refund_output(swap_id, value, secret_hash, receiver));
```

Comment:
Please consider calling payment_status, refund and spend on smart contract address extracted from original swap transaction. The config can change between MM2 restarts which will result in incorrect smart contract address used.

Reply:
Thank you, It's good solution until we have implemented the swap contract address updating.
Done